types/koa-webpack/index.d.ts

```diff
@@ -8,12 +8,20 @@
 // TypeScript Version: 2.3
 
 import Koa = require('koa');
+import MemoryFileSystem = require('memory-fs');
 import webpack = require('webpack');
 import webpackDevMiddleware = require('webpack-dev-middleware');
 import webpackHotClient = require('webpack-hot-client');
 
+declare module 'koa' {
+    interface DefaultState {
```

Comment:
May I ask what would be the use case of these default state fields?
Because I couldn't find `memory-fs` in koa-webpack's dependency list.
If you need to access these fields, this way might work without modifying type file:
```typescript
const app = new Koa<{
    fs: MemoryFileSystem;
    stats: webpack.Stats;
}>();

app.use(ctx => {
    // $ExpectType MemoryFileSystem
    ctx.state.fs;
    ctx.body = ctx.state.fs.createReadStream('file.js');
});

app.use(ctx => {
    // $ExpectType Stats
    ctx.state.stats;
    ctx.body = ctx.state.stats.toJson();
});
```

The README of `memory-fs` package shows that it has been deprecated as well.
We should avoid adding a deprecated package to the dependency list.

Reply:
This is how `DefaultState` is supposed to be used according to the tsdoc ([link](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/koa/index.d.ts#L520-L523)). The same goes for `DefaultContext`.

There are multiple packages on DefinitelyTyped that augment these types this way.

Reply:
Thanks for your explanation :smile:
I'm looking for a clean solution to let TypeScript infer the type itself as much as possible.
But the `compose` situation you are facing seems not easy to infer correctly.
There might be no way to infer the injected values (ctx.state.`fs`) before the app really applies middleware.
The second element of `compose` function can only get the original type of `ctx.state` without `fs`.
I agree with you that `DefaultState` would be the straightforward solution.

Besides, I found that `webpack-dev-middleware` injected the `fs` through these lines:
https://github.com/webpack/webpack-dev-middleware/blob/v3.7.2/lib/fs.js#L138
https://github.com/webpack/webpack-dev-middleware/blob/v3.7.2/lib/middleware.js#L38
To point out where does the `fs` come from.
I think that `webpackDevMiddleware .Options['fs']` is more clear than `MemoryFileSystem`.

I'll review the changes soon so that we can merge it :+1:

Reply:
I agree Koa types are a bit odd.

I changed the type from `MemoryFileSystem` to `webpackDevMiddleware.Options['fs']`